File: src/main/scala/avalon/extend/RSSParser.scala
package avalon.extend

import java.net.URL
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util

import org.apache.commons.lang3.builder.HashCodeBuilder
import org.jdom2.Element
import org.jdom2.input.SAXBuilder

object RSSParser {

	case class RSSItem(title: String, link: String, pubDate: LocalDateTime, info: RSSInfo) {
		override def toString: String = s"{RSSItem: title=$title, link=$link, pubDate=$pubDate, info=$info"

		override def hashCode(): Int = new HashCodeBuilder(37, 17).append(title).append(info.hashCode()).hashCode()
	}

	case class RSSInfo(title: String, link: String, description: String) {
		override def toString: String = s"{RSSInfo: title=$title, link=$link, description=$description"

		override def hashCode(): Int = new HashCodeBuilder(37, 17).append(title).append(link).hashCode()
	}

	def get(url: URL): util.List[RSSItem] = {
		val sb = new SAXBuilder
		val doc = sb.build(url)
		val root = doc.getRootElement.getChild("channel")
		val resultItem = new util.ArrayList[RSSItem]()
		val items = root.getChildren("item").toArray()
		val info = RSSInfo(root.getChildTextNormalize("title"),
			root.getChildTextNormalize("link"),
			root.getChildTextNormalize("description"))
		for (item <- items) {
			val element = item.asInstanceOf[Element]
			resultItem add
				RSSItem(element.getChildTextNormalize("title"),
					element.getChildTextNormalize("link"),
					LocalDateTime.parse(element.getChildTextNormalize("pubDate"), DateTimeFormatter.RFC_1123_DATE_TIME),
					info)
		}
		resultItem
	}
}

File: src/main/kotlin/avalon/extend/RSSParser.kt
package avalon.extend

import org.apache.commons.lang3.builder.EqualsBuilder
import org.apache.commons.lang3.builder.HashCodeBuilder
import org.jdom2.Element
import org.jdom2.input.SAXBuilder
import java.net.URL
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * Created by Eldath Ray on 2017/10/5 0021.
 *
 * @author Eldath Ray
 */
object RSSParser {
	class RSSItem(val title: String, val link: String, val pubDate: LocalDateTime, val info: RSSInfo) {
		override fun toString(): String = "{RSSItem: title=$title, link=$link, pubDate=$pubDate, info=$info"

		override fun hashCode(): Int = HashCodeBuilder(37, 17).append(title).append(info.hashCode()).hashCode()
		override fun equals(other: Any?): Boolean {
			if (this === other) return true
			if (javaClass != other?.javaClass) return false
			other as RSSItem
			return EqualsBuilder().append(other.title, title).append(other.link, link).isEquals
		}
	}

	class RSSInfo(val title: String, val link: String, val description: String) {
		override fun toString(): String = "{RSSInfo: title=$title, link=$link, description=$description"

		override fun hashCode(): Int = HashCodeBuilder(37, 17).append(title).append(link).hashCode()
		override fun equals(other: Any?): Boolean {
			if (this === other) return true
			if (javaClass != other?.javaClass) return false
			other as RSSInfo
			return EqualsBuilder().append(other.title, title).append(other.link, link).isEquals
		}
	}

	@JvmStatic
	fun get(url: URL): List<RSSItem> {
		val sb = SAXBuilder()
		val doc = sb.build(url)
		val root = doc.rootElement.getChild("channel")
		val items = root.getChildren("item").toList()
		val info = RSSInfo(root.getChildTextNormalize("title"),
				root.getChildTextNormalize("link"),
				root.getChildTextNormalize("description"))
		return items.map { it as Element }
				.map {
					RSSItem(it.getChildTextNormalize("title"),
							it.getChildTextNormalize("link"),
							LocalDateTime.parse(it.getChildTextNormalize("pubDate"), DateTimeFormatter.RFC_1123_DATE_TIME),
							info)
				}
	}

}